source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -272,111 +272,183 @@ absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& respon
+         callback_state_ == CallbackState::BufferedPartialBodyCallback;
+}
+
+absl::Status ProcessorState::processResponseBasedOnState(const CommonResponse& common_response,
+                                                         bool& should_continue) {
+  switch (callback_state_) {
+  case CallbackState::BufferedBodyCallback:
+    return handleBufferedBodyCallback(common_response, should_continue);
+  case CallbackState::StreamedBodyCallback:
+    return handleStreamedBodyCallback(common_response, should_continue);
+  case CallbackState::BufferedPartialBodyCallback:
+    return handleBufferedPartialBodyCallback(common_response, should_continue);
+  default:
+    // Fake a grpc error when processor state and received message type doesn't match
+    onFinishProcessorCall(Grpc::Status::FailedPrecondition);
+    return absl::OkStatus();
+  }
+}
+
+absl::Status ProcessorState::handleBufferedBodyCallback(const CommonResponse& common_response,
+                                                        bool& should_continue) {
+  // Handle header mutations if present
+  if (common_response.has_header_mutation()) {
+    const auto mut_status = processHeaderMutationIfAvailable(common_response);
+    if (!mut_status.ok()) {
+      return mut_status;
+    }
+  }
 
-    // Send trailers if they are available and no data pending for processing.
-    if (send_trailers_ && trailers_available_ && chunk_queue_.empty()) {
-      filter_.sendTrailers(*this, *trailers_);
-      return absl::OkStatus();
+  // Handle body mutations if present
+  if (common_response.has_body_mutation()) {
+    const auto validation_status = validateContentLength(common_response);
+    if (!validation_status.ok()) {
+      return validation_status;
     }
+    applyBufferedBodyMutation(common_response);
+  }
+
+  clearWatermark();
+  onFinishProcessorCall(Grpc::Status::Ok);
+  should_continue = true;
+  return absl::OkStatus();
+}
+
+absl::Status ProcessorState::handleStreamedBodyCallback(const CommonResponse& common_response,
+                                                        bool& should_continue) {
+  absl::StatusOr<bool> result = handleBodyInStreamedState(common_response);
+  if (!result.ok()) {
+    return result.status();
+  }
+  should_continue = *result;
+  return absl::OkStatus();
+}
+
+absl::Status
+ProcessorState::handleBufferedPartialBodyCallback(const CommonResponse& common_response,
+                                                  bool& should_continue) {
+  Buffer::OwnedImpl chunk_data;
+  auto chunk = dequeueStreamingChunk(chunk_data);
+  if (!chunk) {
+    ENVOY_BUG(false, "Bad partial body callback state");
+    return absl::InternalError("Invalid chunk in partial body callback");
+  }
 
-    if (should_continue || (trailers_available_ && chunk_queue_.empty())) {
-      continueIfNecessary();
+  // Process header mutations if present
+  if (common_response.has_header_mutation()) {
+    const auto mut_status = processHeaderMutationIfAvailable(common_response);
```

Comment:
can you spell out `mut_status` more fully? I don't know what `mut` is short for.